scipy/spatial/transform/_rigid_transform_xp.py

```diff
@@ -0,0 +1,350 @@
+from types import EllipsisType
+
+from scipy._lib._array_api import array_namespace, Array, is_lazy_array, xp_vector_norm
+import scipy._lib.array_api_extra as xpx
+from scipy.spatial.transform._rotation_xp import (
+    as_matrix as quat_as_matrix,
+    from_matrix as quat_from_matrix,
+    from_rotvec as quat_from_rotvec,
+    as_rotvec as quat_as_rotvec,
+    broadcastable,
+    compose_quat,
+    from_quat,
+    inv as quat_inv,
+)
+from scipy._lib.array_api_compat import device
+
+
+def from_matrix(matrix: Array, normalize: bool = True, copy: bool = True) -> Array:
+    xp = array_namespace(matrix)
+    # Shape check should be done before calling this function
+    if normalize or copy:
+        matrix = xp.asarray(matrix, copy=True)
+
+    last_row_ok = xp.all(
+        matrix[..., 3, :] == xp.asarray([0, 0, 0, 1.0], device=device(matrix)), axis=-1
+    )
+    if is_lazy_array(matrix):
+        matrix = xp.where(last_row_ok[..., None, None], matrix, xp.nan)
```

Comment:
To my limited understanding, the input can be integer.
This causes lazy backends to automatically promote everything to float64, if they support int to float promotion (which is undefined behaviour in the Standard).
```suggestion
        matrix = xp_promote(matrix, force_floating=True)
        matrix = xp.where(last_row_ok[..., None, None], matrix, xp.nan)
```
Is this behaviour documented? Is it tested?